tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleFeatureContext.php

```diff
@@ -0,0 +1,354 @@
+    {
+        $data = $this->localizeByRows($node);
+        $defaultLanguageId = Configuration::get('PS_LANG_DEFAULT');
+
+        $cartRuleAction = $this->createCartRuleAction(
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['free_shipping']),
+            $data['reduction_percentage'] ?? null,
+            isset($data['reduction_apply_to_discounted_products']) ? PrimitiveUtils::castStringBooleanIntoBoolean($data['reduction_apply_to_discounted_products']) : null,
+            $data['reduction_amount'] ?? null,
+            isset($data['reduction_currency']) ? $this->getSharedStorage()->get($data['reduction_currency']) : null,
+            $data['reduction_tax'] ?? null,
+            $data['gift_product_id'] ?? null,
+            $data['gift_product_attribute_id'] ?? null
+        );
+
+        if (isset($data['name'])) {
+            $name = $data['name'];
+        } else {
+            //@todo: remove this fallback and change all scenarios to support multilang instead
+            $name = [$defaultLanguageId => $data['name_in_default_language']];
+        }
+
+        $command = new AddCartRuleCommand(
+            $name,
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['highlight']),
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['allow_partial_use']),
+            (int) $data['priority'],
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['is_active']),
+            new DateTime($data['valid_from']),
+            new DateTime($data['valid_to']),
+            $data['total_quantity'],
+            $data['quantity_per_user'],
+            $cartRuleAction
+        );
+
+        if (!empty($data['minimum_amount'])) {
+            $currencyId = SharedStorage::getStorage()->get($data['minimum_amount_currency']);
+            $command->setMinimumAmount(
+                $data['minimum_amount'],
+                $currencyId,
+                PrimitiveUtils::castStringBooleanIntoBoolean($data['minimum_amount_tax_included']),
+                PrimitiveUtils::castStringBooleanIntoBoolean($data['minimum_amount_shipping_included'])
+            );
+        }
+
+        $command->setDescription($data['description'] ?? '');
+        $command->setCode($data['code'] ?? '');
+
+        if (isset($data['discount_application_type'])) {
+            $command->setDiscountApplication(
+                $data['discount_application_type'],
+                // if specific product type is provided and product is not, then command should throw exception
+                isset($data['discount_product']) ? $this->getSharedStorage()->get($data['discount_product']) : null
+            );
+        }
+
+        /** @var CartRuleId $cartRule */
+        $cartRule = $this->getCommandBus()->handle($command);
+
+        //@todo: should be refactored to save only id to shared storage instead of whole object model
+        SharedStorage::getStorage()->set($cartRuleReference, new CartRule($cartRule->getValue()));
```

Comment:
```suggestion
        $this->getSharedStorage()->set($cartRuleReference, new CartRule($cartRule->getValue()));
```
(this should be replaced in the whole file)

Reply:
We should stop using the singleton method, maybe it would be simpler to actually remove it completely

Reply:
However I don't think the refacto should be done in another PR, all our references are based on IDs and we should always do it this way. It forces fetching the object instead of relying on an existing object from previous step